test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,316 @@
+   almost normally.  But instead of running all the controllers concurrently,
+   with their informers connected to their workers through their workqueues,
+   the informers and workers are decoupled.  The test synchronously invokes
+   the workers one at a time.  Before advancing to the next iteration,
+   the test waits in real time for each informer to deliver notifications
+   of what the workers wrote.  The workers read a fake clock, so that the
+   test can exactly control how that time advances.  After every worker run,
+   the test evaluates whether that worker has been sufficiently reticent
+   about updating FlowSchemaStatus.
+
+   This test disables the usual controller in the kube-apiserver so that
+   it does not modify FlowSchemaStatus; such modifications could confuse
+   the test logic that waits for notifications of the ResourceVersions
+   created by the test workers.
+*/
+type fightTest struct {
+	t              *testing.T
+	ctx            context.Context
+	loopbackConfig *rest.Config
+	size           int
+	stopCh         chan struct{}
+	now            time.Time
+	clk            *clock.FakeClock
+	ctlrs          map[bool][]utilfc.TestableInterface
+
+	// lastRVs maps FlowSchema name to the ResourceVersion
+	// most recently created by a controller.
+	lastRVs map[string]string
+
+	rvMutex sync.Mutex // hold when accessing notifiedRVs
+
+	// notifiedRVs maps invert -> i -> FS name -> last notified ResourceVersion
+	notifiedRVs map[bool][]map[string]string
+
+	someTimedOut bool
+	iteration    int
+	// maps invert -> i -> FS name -> write times (oldest first)
+	writeHistories map[bool][]map[string][]time.Time
+}
+
+func newFightTest(t *testing.T, loopbackConfig *rest.Config, size int) *fightTest {
+	now := time.Now()
+	ft := &fightTest{
+		t:              t,
+		ctx:            context.Background(),
+		loopbackConfig: loopbackConfig,
+		size:           size,
+		stopCh:         make(chan struct{}),
+		now:            now,
+		clk:            clock.NewFakeClock(now),
+		ctlrs: map[bool][]utilfc.TestableInterface{
+			false: make([]utilfc.TestableInterface, size),
+			true:  make([]utilfc.TestableInterface, size)},
+		notifiedRVs: map[bool][]map[string]string{
+			false: make([]map[string]string, size),
+			true:  make([]map[string]string, size)},
+		writeHistories: map[bool][]map[string][]time.Time{
+			false: make([]map[string][]time.Time, size),
+			true:  make([]map[string][]time.Time, size)},
+	}
+	ft.foreach(func(invert bool, i int) {
+		ft.rvMutex.Lock()
+		defer ft.rvMutex.Unlock()
+		ft.notifiedRVs[invert][i] = map[string]string{}
+		ft.writeHistories[invert][i] = map[string][]time.Time{}
+	})
+	return ft
+}
+
+func (ft *fightTest) createController(invert bool, i int) {
+	myConfig := rest.CopyConfig(ft.loopbackConfig)
+	myConfig = rest.AddUserAgent(myConfig, fmt.Sprintf("invert=%v, i=%d", invert, i))
+	myClientset := clientset.NewForConfigOrDie(myConfig)
+	fcIfc := myClientset.FlowcontrolV1beta1()
+	if ft.lastRVs == nil {
```

Comment:
Can we move this initialization out of createController?

It sounds like it should be one before we start creating controllers in the test itself.

Reply:
The current logic _does_ initialize `lastRVs` before creating any actual controllers.  I originally considered doing that initialization before the loop, but that requires duplicating the code to create the config, tweak the config, create the clientset, and navigate to the operations interface.